pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/SanitizationTransformer.java

```diff
@@ -36,47 +37,108 @@
 
   public SanitizationTransformer(Schema schema) {
     for (FieldSpec fieldSpec : schema.getAllFieldSpecs()) {
       if (!fieldSpec.isVirtualColumn() && fieldSpec.getDataType() == DataType.STRING) {
-        _stringColumnMaxLengthMap.put(fieldSpec.getName(), fieldSpec.getMaxLength());
+        _stringColumnToFieldSpecMap.put(fieldSpec.getName(), fieldSpec);
       }
     }
   }
 
   @Override
   public boolean isNoOp() {
-    return _stringColumnMaxLengthMap.isEmpty();
+    return _stringColumnToFieldSpecMap.isEmpty();
   }
 
   @Override
   public GenericRow transform(GenericRow record) {
-    for (Map.Entry<String, Integer> entry : _stringColumnMaxLengthMap.entrySet()) {
+    for (Map.Entry<String, FieldSpec> entry : _stringColumnToFieldSpecMap.entrySet()) {
       String stringColumn = entry.getKey();
-      int maxLength = entry.getValue();
       Object value = record.getValue(stringColumn);
       if (value instanceof String) {
         // Single-valued column
-        String stringValue = (String) value;
-        String sanitizedValue = StringUtil.sanitizeStringValue(stringValue, maxLength);
-        // NOTE: reference comparison
-        //noinspection StringEquality
-        if (sanitizedValue != stringValue) {
-          record.putValue(stringColumn, sanitizedValue);
+        Pair<String, Boolean> result = sanitizeValue(stringColumn, (String) value, entry.getValue());
+        record.putValue(stringColumn, result.getLeft());
+        if (result.getRight()) {
+          record.putValue(GenericRow.INCOMPLETE_RECORD_KEY, true);
         }
       } else {
         // Multi-valued column
         Object[] values = (Object[]) value;
-        int numValues = values.length;
-        for (int i = 0; i < numValues; i++) {
-          values[i] = StringUtil.sanitizeStringValue(values[i].toString(), maxLength);
+        for (int i = 0; i < values.length; i++) {
+          Pair<String, Boolean> result = sanitizeValue(stringColumn, values[i].toString(), entry.getValue());
+          values[i] = result.getLeft();
+          if (result.getRight()) {
+            record.putValue(GenericRow.INCOMPLETE_RECORD_KEY, true);
+          }
         }
       }
     }
     return record;
   }
+
+  /**
+   * Sanitize the value for the given column. Usually a STRING column can be extended for JSON / BYTES in future.
+   * @param stringColumn column name
+   * @param value value of the column
+   * @param columnFieldSpec field spec of the column defined in schema
+   * @return the sanitized value and a boolean indicating if the value was sanitized
+   */
+  private Pair<String, Boolean> sanitizeValue(String stringColumn, String value, FieldSpec columnFieldSpec) {
+    String sanitizedValue = StringUtil.sanitizeStringValue(value, columnFieldSpec.getMaxLength());
+    FieldSpec.MaxLengthExceedStrategy maxLengthExceedStrategy;
+    if (columnFieldSpec.getMaxLengthExceedStrategy() == null) {
```

Comment:
Suggest doing this check in the constructor instead of on each row.
We'll need maxLength, maxLengthExceedStrategy, defaultValue for each column. We can make a private class to hold them